src/Passes/CONTRIBUTING.md

```diff
@@ -0,0 +1,83 @@
+
+| **Code Element**      | **Style**                        | **Comment**                                                                                                                                   |
+| --------------------- | -------------------------------- | --------------------------------------------------------------------------------------------------------------------------------------------- |
+| Namespace             | snake_case                       | Differentiates `namespace::ClassName` and `ClassName::SubClass` names                                                                         |
+| Class name            | CamelCase                        | To differentiate from STL types which ISO recommends (do not use "C" or "T" prefixes)                                                         |
+| Function name         | camelCase                        | Lower case start is almost universal except for .Net world                                                                                    |
+| Parameters/Locals     | snake_case                       | Vast majority of standards recommends this because \_ is more readable to C++ crowd (although not much to Java/.Net crowd)                    |
+| Member variables      | snake_case_with\_                | The prefix \_ is heavily discouraged as ISO has rules around reserving \_identifiers, so we recommend suffix instead                          |
+| Enums and its members | CamelCase                        | Most except very old standards agree with this one                                                                                            |
+| Globals               | g_snake_case                     | Avoid using globals whenever possible, but if you have to use `g_`.                                                                           |
+| Constants             | UPPER_CASE                       | Very contentious and we just have to pick one here, unless if is a private constant in class or method, then use naming for Members or Locals |
+| File names            | Match case of class name in file | Lot of pro and cons either way but this removes inconsistency in auto generated code (important for ROS)                                      |
+
+## Modernise when possible
+
+In general, modernise the code where possible. For instance, prefer `using` of `typedef`.
+
+## Header guards
+
+Prefer `#pragma once` over `#ifdef` protection.
+
+## Code TODOs must contain owner name or Github issue
+
+```sh
+% ./manage runci
+(...)
+QsPasses/src/OpsCounter/OpsCounter.cpp:39:21: error: missing username/bug in TODO [google-readability-todo,-warnings-as-errors]
+                    // TODO: Fails to load if this is present
+                    ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+                    // TODO(tfr): Fails to load if this is present
+```
+
+## Always add copyrights
+
+Always add copyrights at the top of the file.
```

Comment:
(I'm not strong here, more opinions are recommended)
Consider either excluding this section or something like
> Add MS copyrights at the top of the newly-created files and files consisting of fragments from MS-owned sources. For non-MS-owned sources consider adding those sources without changes and preserving their copyrights.

(And I would request specifically for more opinions about the case when a non-MS employee (or even the whole company) adds a new file, what copyright should be used there, if any)

Reply:
The more important part is that if you _do_ choose to add a copyright it should be at the top of the file (and not in bottom or in the middle of the code). It is about consistency more than it is about MS copyright practices. 

Right now, the scripts are set up to enforce a very strict requirement that all files must have MS copyright notice at the top. We should probably lose that restriction as I didn't really consider the case of non-MS copyrights. That is definitely a case that we also need to deal with. I am happy to remove this section.  Let me know your thoughts.

A separate, but important part of the discussion is what the copyright notice should look like. I understand that we may want to move this part of the code into a collaborative repository which will be built with our partners. In this case, it may be more appropriate to have a copyright statement along the lines of 
```
// Copyright (c) The __PROJECT NAME__ Authors
// Licensed under the MIT License.
```
with a `AUTHORS` file at the root of the repository. This removes MS as being the perceived dominant copyright holder to more of a collaboration effort. This practice is for instance used in the `Chromium` project. @bettinaheim what are your thoughts?

A third discussion to be had: Do we allow copy/paste of external code (if appropriate copyright is preserved)? My hunch is that this should only be allowed (if at all) if it is not possible to include it as a library.

Note, I have no strong preference for either of these issues, the only thing that is important for me is that we have some level of consistency across files.

Reply:
"Right now, the scripts are set up to enforce a very strict requirement that all files must have MS copyright notice at the top. " That is indeed the requirement we have to stick to as long as the code lives on this repo. A chance in that will be done (only) as part of moving (contributing/donating) the code in the future. Please do not add any instructions on the copyright to the readme beyond maybe stating that copy right headers go at the top, since guidance on the copy right may change and it's our job upon code review to make sure the correct copy right is there.

Reply:
Marking as resolved.